src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverTests_Attributes_FullyQualifiedName.cs

```diff
@@ -0,0 +1,1113 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Test.Utilities.TestGenerators;
+using Roslyn.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.Semantic.UnitTests.SourceGeneration;
+
+internal static class IncrementalGeneratorInitializationContextExtensions
+{
+    public static IncrementalValuesProvider<T> ForAttributeWithSimpleName<T>(
+        this IncrementalGeneratorInitializationContext context, string simpleName)
+        where T : SyntaxNode
+    {
+        return context.ForAttributeWithSimpleName(
+            simpleName,
+            (node, _) => node is T).Select((node, _) => (T)node);
+    }
+
+    public static IncrementalValuesProvider<T> ForAttributeWithMetadataName<T>(
+        this IncrementalGeneratorInitializationContext context, string fullyQualifiedMetadataName)
+        where T : SyntaxNode
+    {
+        return context.ForAttributeWithMetadataName(
+            fullyQualifiedMetadataName,
+            (node, _) => node is T,
+            (context, cancellationToken) => (T)context.AttributeTarget);
+    }
+}
+
+public class GeneratorDriverTests_Attributes_FullyQualifiedName : CSharpTestBase
+{
+    #region Non-Incremental tests
+
+    // These tests just validate basic correctness of results in different scenarios, without actually validating
+    // that the incremental nature of this provider works properly.
+
+    [Fact]
```

Comment:
Did you consider making this a `[Theory]` where you could abstract over the various type kinds? Essentially 

```
[Theory]
[InlineData("class")]
[InlineData("struct")]
...
public void FindCorrectAttributeOnTopLevelClass_WhenSearchingForClassDeclaration1(string declKind)
{
    {
        var source = $$"""
[N1.X]
{{declKind}} C1 { }
[N2.X]
{{declKind}} C2 { }
""";
```

The use of `ClassDeclarationSyntax` could be replaced with `TypeDeclarationSyntax`

Reply:
i didn't feel that there was *huge* value in testing that.  i do have tests for the various types of attributable nodes that are *not* type declarations.  overall i thought that was sufficient :)